core/services/head_tracker.go

```diff
@@ -391,59 +387,61 @@ func (ht *HeadTracker) backfill(ctx context.Context, head models.Head, baseHeigh
 			head = *existingHead
 			continue
 		}
-		head, err = ht.fetchAndSaveHead(ctx, i)
-		fetched++
-		if err != nil {
-			if errors.Cause(err) == ethereum.NotFound {
-				logger.Errorw("HeadTracker: backfill failed to fetch head (not found), chain will be truncated for this head", "headNum", i)
-			} else if errors.Cause(err) == context.DeadlineExceeded {
-				logger.Warnw("HeadTracker: backfill deadline exceeded, chain will be truncated for this head", "headNum", i)
-			} else {
-				logger.Errorw("HeadTracker: backfill encountered unknown error, chain will be truncated for this head", "headNum", i, "err", err)
-			}
-			break
-		}
+		headNumbers = append(headNumbers, big.NewInt(i))
 	}
-	return nil
-}
 
-func (ht *HeadTracker) fetchAndSaveHead(ctx context.Context, n int64) (models.Head, error) {
-	logger.Debugw("HeadTracker: fetching head", "blockHeight", n)
-	head, err := ht.fetchHead(ctx, n)
+	maybeHeads, err := ht.store.EthClient.BatchHeaderByNumber(ctx, headNumbers)
 	if err != nil {
-		return head, err
+		ht.handleBackfillError(err, -1)
+		return err
 	}
-	if err := ht.store.IdempotentInsertHead(head); err != nil {
-		return head, err
+
+	for _, maybeHead := range maybeHeads {
+		if maybeHead.Error != nil {
+			ht.handleBackfillError(maybeHead.Error, maybeHead.Number)
+			break
+		} else if maybeHead.Header == nil {
+			logger.Warn("got nil block header")
+			break
+		}
+		gethHeader := maybeHead.Header
+		head := models.NewHead(gethHeader.Number, gethHeader.Hash(), gethHeader.ParentHash, gethHeader.Time)
```

Comment:
This doesnt work with parity.

Reply:
You HAVE to use the RPC call.

Reply:
I swear I added a regression test for this...

Reply:
Fixed